### PR TITLE
Add support for maintaining groups that are shared with other groups

### LINKED ISSUE #150: Enforce group sharing
I need to set up a group to be shared with another group. For example, a groupP1 containing projects will be shared with groupU1. This seems doable but if I manually share groupP1 with groupU2, gitlabform doesn't remove that manual share even if I have the enforcement option enabled. I think this is because the enforcement option currently enforces members but not the shared groups.

Can the permission enforcement option for groups be enhanced so that group sharing is enforced also?

### PR DESCRIPTION
**Pull request by andrewjw**
Hi,

This PR adds a new `group_shared_with` group level config, and code to maintain the list of groups that a group is shared with. It also includes acceptance tests, which pass (at least for me). The code is very much based on the code that maintains group user access.

I believe this fixes #150.

Please review the code, I'm happy to make any changes that might be required.

Many thanks,
Andrew Wilkinson (on behalf of Ocado Technology)

### CODE REVIEW

File: gitlabform/gitlabform/processors/group/group_shared_with_processor.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+        groups_to_set_by_group_id = set()
+
+        for groupname in groups_to_set_by_groupname:
+
+            try:
+                group_id = self.gitlab.get_group_case_insensitive(groupname)["id"]
+            except NotFoundException:
+                cli_ui.error(f"Group {groupname} not found.")
+                sys.exit(EXIT_INVALID_INPUT)
```
Comment (gdubicki) [inline gitlabform/gitlabform/processors/group/group_shared_with_processor.py:40]: The pattern in our current codebase for this is to convert names to ids inside the `gitlabform.gitlab` package, see https://github.com/egnyte/gitlabform/blob/d926f5561007d82c88e1a242c927993527a61c9f/gitlabform/gitlab/projects.py#L229-L236 as an example. I think that we should do the same thing here.

Comment (gdubicki) [reply gitlabform/gitlabform/processors/group/group_shared_with_processor.py:40]: I guess that this is done this way for performance reasons, to not re-do the same get group id requests over and over, and that makes sense, but I propose to cover this by introducing cache to wrap around some methods that we know that are safe to cache, such as those getting object ids.

Comment (gdubicki) [reply gitlabform/gitlabform/processors/group/group_shared_with_processor.py:40]: WDYT?

File: gitlabform/gitlabform/test/test_group_shared_with.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            projects_and_groups:
+              {group}/*:
+                group_members:
+                  {users[0]}:
+                    access_level: 50
+                group_shared_with:
+                  {groups[0]}:
+                    group_access_level: 30
+            """,
+        ]
```
Comment (gdubicki) [inline gitlabform/gitlabform/test/test_group_shared_with.py:111]: I think that these 2 examples would not remove the sharing even with `enforce_group_members: true`. Isn't the `group_shared_with` content in them exactly the same as at the end of the previous test:
```
            group_shared_with:
              {groups[0]}:
                group_access_level: 30
```
?

**Review by gdubicki — CHANGES_REQUESTED** [review]
Hey @andrewjw! Thank you very much for your contribution! Please check out my review.

File: gitlabform/gitlabform/processors/group/group_shared_with_processor.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+
+class GroupSharedWithProcessor(AbstractProcessor):
+    def __init__(self, gitlab: GitLab):
+        super().__init__("group_shared_with")
+        self.gitlab = gitlab
+
+    def _process_configuration(self, group: str, configuration: dict):
+        groups_to_set_by_groupname = configuration.get("group_shared_with")
+        print(groups_to_set_by_groupname)
```
[the remark this answers is not in the document]
Comment (andrewjw) [reply gitlabform/gitlabform/processors/group/group_shared_with_processor.py]: Sorry, I missed a couple. They've been removed now.

File: gitlabform/gitlabform/processors/group/group_shared_with_processor.py (hunk 1 above)
Comment (andrewjw) [reply gitlabform/gitlabform/processors/group/group_shared_with_processor.py]: Yes, I was worried about the performance of many requests to convert the group name into an id. I've changed this now so the api uses group names, and internally caches the mapping to an id. Please let me know if this looks ok now.

File: gitlabform/gitlabform/test/test_group_shared_with.py (hunk 2 above)
Comment (andrewjw) [reply gitlabform/gitlabform/test/test_group_shared_with.py]: Thanks, good spot, I've corrected this.

File: gitlabform/gitlab/groups.py [hunk 4]
```diff
@@ -1,7 +1,18 @@
...
+import sys
+
+import cli_ui
+
+from gitlabform import EXIT_INVALID_INPUT
 from gitlabform.gitlab.core import GitLabCore, NotFoundException
 
 
 class GitLabGroups(GitLabCore):
+    @functools.lru_cache
```
Comment (gdubicki) [inline gitlabform/gitlab/groups.py:11]: Check out the failed tests, @andrewjw . 
You need to add `()` for this code to work in Python 3.6 and 3.7.
Only in 3.8+ you can omit the parentheses if you don't pass any parameters.
(Btw: I learned that today. 😅 )

Comment (andrewjw) [reply gitlabform/gitlab/groups.py]: I have to admit I only tested it on Python 3.8. Hopefully all the tests will pass now.

**Comment by gdubicki** [comment]
Thank you again, @andrewjw ! Please check out the new v2.1.0 release with your new feature.

**Pull request merged**